data/account/participationRegistry.go

```diff
@@ -332,61 +415,176 @@ func (db *participationDB) GetAll() (records []ParticipationRecord, err error) {
 	return
 }
 
+func (db *participationDB) Get(id ParticipationID) (record ParticipationRecord, err error) {
+	db.mutex.RLock()
+	defer db.mutex.RUnlock()
+
+	record, ok := db.cache[id]
+	if !ok {
+		return ParticipationRecord{}, ErrParticipationIDNotFound
+	}
+	return record, nil
+}
+
+func (db *participationDB) GetAll() ([]ParticipationRecord, error) {
+	db.mutex.RLock()
+	defer db.mutex.RUnlock()
+
+	results := make([]ParticipationRecord, 0, len(db.cache))
+	for _, record := range db.cache {
+		copy := record
+		results = append(results, copy)
+	}
+	return results, nil
+}
+
+// updateRollingFields sets all of the rolling fields according to the record object.
+func (db *participationDB) updateRollingFields(ctx context.Context, tx *sql.Tx, record ParticipationRecord) error {
+	result, err := tx.ExecContext(ctx, updateRollingFields,
+		record.LastVote,
+		record.LastBlockProposal,
+		record.LastCompactCertificate,
+		record.RegisteredFirst,
+		record.RegisteredLast,
+		record.ParticipationID[:])
+	if err != nil {
+		return err
+	}
+
+	numRows, err := result.RowsAffected()
+	if err != nil {
+		return err
+	}
+	if numRows > 1 {
```

Comment:
What about numRows==0 ?

Reply:
Thanks, I'll return an error here. I was thinking that it's safe to ignore, but looking closer that isn't always true - specifically when setting that a key is registered.

Reply:
This thing is - if the key is missing from the database, and we have the entry in the cache, we might want to report the error, but then - also remove it from the cache to avoid repeating this error.
( since the database won't "fix" itself, unless we somehow "fix" the situation )